src/js/AccessibilityUtils.js

```diff
@@ -1034,29 +1034,57 @@ axs.utils.isInlineElement = function(element) {
 };
 
 /**
- * @param {Element} element
- * @return {Object|boolean}
+ *
+ * Gets role details from an element.
+ * @param {Element} element The DOM element whose role we want.
+ * @param {boolean=} implicit if true then implicit semantics will be considered if there is no role attribute.
+ *
+ * @return {Object}
  */
-axs.utils.getRoles = function(element) {
-    if (!element.hasAttribute('role'))
-        return false;
+axs.utils.getRoles = function(element, implicit) {
+    if (!element || element.nodeType !== Node.ELEMENT_NODE || (!element.hasAttribute('role') && !implicit))
+        return null;
     var roleValue = element.getAttribute('role');
+    if (!roleValue && implicit)
+        roleValue = axs.properties.getImplicitRole(element);
+    if (!roleValue)  // role='' or implicit role came up empty
+        return null;
     var roleNames = roleValue.split(' ');
     var roles = [];
     var valid = true;
     for (var i = 0; i < roleNames.length; i++) {
         var role = roleNames[i];
-        if (axs.constants.ARIA_ROLES[role])
-            roles.push({'name': role, 'details': axs.constants.ARIA_ROLES[role], 'valid': true});
-        else {
-            roles.push({'name': role, 'valid': false});
+        var ariaRole = axs.constants.ARIA_ROLES[role];
+        if (ariaRole && !ariaRole.abstract) {
+            var roleObject = {'name': role, 'details': axs.constants.ARIA_ROLES[role], 'valid': true};
+            roles.push(roleObject);
+        } else {
+            var roleObject = {'name': role, 'valid': false};
             valid = false;
+            roles.push(roleObject);
         }
     }
 
     return { 'roles': roles, 'valid': valid };
 };
 
+RoleDetails.prototype.contains = function (role) {
+    var roles = this.roles;
+    if (roles.length) {
+        return (roles.indexOf(role) >= 0);
+    }
+    return false;
+};
+
+/**
+ * Represents 
```

Comment:
Looks a bit unfinished... it doesn't look like this type is used, either, so maybe scrap it for this change?